File: day-2/part-2/elixir/coordinates.exs
defmodule Coordinates do

  def calculate(input) do
    directions = prepare_directions(input)
    {x, y, _} = directions
      |> Enum.reduce({0, 0, 0}, fn direction, acc ->
        move(direction, acc)
      end)
    x * y
  end

  defp prepare_directions(input) when is_binary(input) do
    input
      |> String.split("\n")
      |> Enum.map(&String.trim/1)
      |> Enum.filter(&(String.length(&1) > 0))
  end

  defp prepare_directions(file) do
    case file do
      {:ok, content} -> prepare_directions(content)
      {:error, _} -> []
    end
  end

  defp move(<<"forward ", number::binary>>, {x, y, aim}) do
    {number, _} = Integer.parse(number)
    {x + number, y + (aim * number), aim}
  end

  defp move(<<"up ", number::binary>>, {x, y, aim}) do
    {number, _} = Integer.parse(number)
    {x, y, aim - number}
  end

  defp move(<<"down ", number::binary>>, {x, y, aim}) do
    {number, _} = Integer.parse(number)
    {x, y, aim + number}
  end
end

File: day-2/part-2/ruby/coordinates.rb
class Coordinates
  def calculate(input)
    directions = prepare_directions(input)
    coordinates = directions.reduce({x: 0, y: 0, aim: 0}) do |acc, direction|
      move(direction, acc)
    end
    coordinates[:x] * coordinates[:y]
  end

  private

  def move(direction, position)
    position => {x:, y:, aim:}
    cor, number = direction.split(' ')
    case cor
    when 'forward'
      {x: x + number.to_i, y: y + (aim * number.to_i), aim: aim}
    when 'up'
      {x: x, y: y, aim: aim - number.to_i}
    when 'down'
      {x: x, y: y, aim: aim + number.to_i}
    end
  end

  def prepare_directions(file)
    file.split("\n")
  end
end